XCOFF implicit.ref metadata

mandlebug opened the pull request:
Add a new metadata node `!implicit.ref` to represent an implicit dependency between 2 symbols. The metadata is unique to AIX and gets lowered to a relocation that adds an explicit link between the section the global that the metadata is placed on is allocated in, to the asscoiated symbol. This relocation will cause the associated symbol to remain live if the section is not garbage collected. This is used mainly for compiler features where there is some hidden runtime dependency between the symbols that isn't otherwise obvious to the linker.

mandlebug committed:
Extend the associated metadata node impl for XCOFF.

Add support for the associated metadata node for AIX.. Map it to
the .ref assembly directive. Placing the node on a symbol will cause
the back end to emit a .ref directive in the section of the global the
metadata is placed on. the .ref will refrence the associated symbol
which will keep it alive if the section is not garbage collected.

mandlebug committed:
Run clang-format over changes.

w2yehia commented on llvm/test/CodeGen/PowerPC/aix-associated-metadata.ll:
```diff
@@ -0,0 +1,23 @@
+; RUN: llc -verify-machineinstrs -mtriple powerpc-ibm-aix-xcoff < %s | FileCheck %s
```
I think we need a test for the object file output

mandlebug replied:
I've added textual assembly testing and started working on adding the matching object generation tests. Do you know if the out of order relocations in an issue for function as well?

w2yehia replied:
> Do you know if the out of order relocations in an issue for function as well?

yes, if you generate a `.ref` in the midst of generating a function body, and -fno-function-sections is ON (so there's just the singleton `.text` csect), the relocation gets generated at offset zero in that section.
Eg.
```
                        ***Relocation Information***
             Vaddr      Symndx  Sign  Fixup     Len      Type  Name
.text:
        0x000000c2  0x00000025     0      0  0x000f TOC_ILodx  _system_configuration
        0x000000d6  0x00000027     0      0  0x000f TOC_ILodx  foo.cpu_pwr9
        0x000000de  0x00000025     0      0  0x000f TOC_ILodx  _system_configuration
        0x000000f2  0x00000029     0      0  0x000f TOC_ILodx  foo.cpu_pwr8
        0x000000fa  0x0000002b     0      0  0x000f TOC_ILodx  foo.default
        0x00000000  0x00000015     0      0  0x0000   Non_Rel  ifunc_sec
        0x00000000  0x00000005     0      0  0x0000   Non_Rel  .__init_ifuncs
        0x0000017e  0x0000002d     0      0  0x000f TOC_ILodx  foo
....
                        ***Symbol Table Information***
[Index] m        Value       Scn   Aux  Sclass    Type            Name
[Index] a0                                                        Fname
[Index] a1      Tagndx      Lnno  Size  Lnoptr    Endndx
[Index] a2      Tagndx            Fsiz  Lnoptr    Endndx
[Index] a3      Tagndx      Lnno  Size  Dimensions
[Index] a4   CSlen     PARMhsh SNhash SMtype SMclass Stab SNstab
[Index] a5      SECTlen    #RELent    #LINnums

[7]     m   0x00000000     .text     1  unamex                    
[8]     a4  0x00000190       0    0     SD       PR    -    -
[9]     m   0x00000000     .text     1  unamex                    .foo.default
[10]    a4  0x00000007       0    0     LD       PR    -    -
[11]    m   0x00000040     .text     1  unamex                    .foo.cpu_pwr9
[12]    a4  0x00000007       0    0     LD       PR    -    -
[13]    m   0x00000080     .text     1  unamex                    .foo.cpu_pwr8
[14]    a4  0x00000007       0    0     LD       PR    -    -
[15]    m   0x000000c0     .text     1  unamex                    .foo.resolver
[16]    a4  0x00000007       0    0     LD       PR    -    -
[17]    m   0x00000120     .text     1  extern                    .main
[18]    a4  0x00000007       0    0     LD       PR    -    -
[19]    m   0x0000017c     .text     1  extern                    .foo
[20]    a4  0x00000007       0    0     LD       PR    -    -       
```
The bug in the object gen is that it puts the relocation at the start of whatever csect the function is in, instead of (I imagine) the current offset. You can probably reproduce with -ffunction-sections if you call emitXCOFFRefDirective while in the middle of a function body generation, and there's been prior relocations emitted in that function.

w2yehia reviewed (approved):
LGTM once comments are addressed.

mandlebug committed:
Test .ref impl with data sections disabled.

mandlebug committed:
Check multiple associated metadata nodes in the verifier.

mandlebug committed:
Add testing of the integrated assembler path.

mandlebug commented on llvm/docs/LangRef.rst:
```diff
@@ -8124,6 +8124,13 @@ See :doc:`CalleeTypeMetadata`.
 a single argument that references a global object (optionally through an alias).
+The metadata is often used with an explicit section consisting of valid C
+identifiers so that the runtime can find the metadata section with
+linker-defined encapsulation symbols ``__start_<section_name>`` and
```
Thanks, that was very informative. I still need to read through the linked 'COMDAT and section group' blog post as its relevant to this work as well. This and the rename metadata together is our solution to scenario 2 you describe in the post (the metadata sections are allocatable and appear unreferenced from the code but will be referenced from the runtime), and sort of replicates section groups IIUC. On AIX garbage collection is enabled by default and code is usually compiled with both data sections and function sections enabled. This can lead to people having builds that need linker garbage collection on and working well to function sometimes.

By naming convention do you mean creating the symbols as `__start_<section_name>` and `__stop_<section_name>`? If so yes it follows the same convention.

MaskRay reviewed (commented):
In ELF, `only a single associated metadata is supported`.
If a global object `linkorder` has an `!associated` referencing `text`, it means that 

1. retaining section(text) should also retain section(linkorder)
2. discarding section(text) should also discard section(linkorder). If `linkorder` is referenced by a relocation from a live section, lld will report `error: a.o:(.linkorder): sh_link points to discarded section a.o:(.text)`

A relocation approach can satisfy case 1. While it cannot detect case 2, the limitation seems acceptable.

The implementation appears to function essentially as an `ELF R_PPC64_NONE` relocation with an inverted relationship. I think this should be fixed.

For functions, if https://github.com/llvm/llvm-project/pull/147427  is accepted, you can lower the `llvm.reloc.none` intrinsic to XCOFF `R_REF` relocation.

mandlebug commented:
Now that I understand `SHF_LINK_ORDER` better I think trying to extend the associated metadata for XCOFF is the wrong approach. The ref relocation is used in the same situation (ie trying to make an explicit link between metadata sections to direct garbage collection) but the inverted relationship compared to SHF_LINK_ORDER, missing the link error if a metadata section is referred to by another section,  the fact that the ref directive is also useful in other contexts , as well as potentially needing more then 1 on a global variable makes it more appropriate to add a new metadata instead. In light of #147427 I think it makes sense to add a new metadata kind that can expand out to the ref relocation on XCOFF and the NONE relocation on ELF. That way associated keeps its very specific semantics it has now, and we have a mechanism for emitting these REF/NONE relocations either on functions or data.

mandlebug committed:
Add a new metadata node rather then piggybacking on associated.

w2yehia commented on llvm/docs/LangRef.rst:
```diff
@@ -8440,6 +8440,35 @@ Example:
+.ref directive which will emit a relocation introducing an explicit dependence
+to the referenced symbol. This is typically used when there is some implicit
+dependence between the symbols that is otherwise opaque to the linker. One such
+example is metadata which is accessed by a runtime with associated
```
the term metadata is vague, maybe say:
One such example is "metadata" variables, associated to a function, that are accessed by a runtime via __start_<section_name> and __stop_<section_name> symbols. The function would have a `!ref` MD referencing the variables.

w2yehia commented on llvm/docs/LangRef.rst:
```diff
@@ -8440,6 +8440,35 @@ Example:
+example is metadata which is accessed by a runtime with associated
+``__start_<section_name>`` and ``__stop_<section_name>`` symbols.
+
+This metadata lowers to the .ref assembly directive which will add a relocation
```
We already described what the MD is lowered too.

MaskRay commented:
Using a different metadata is the right direction. While the implementation looks fine, I am unsure whether a simple word `ref` is the best way to name this not-commonly-used feature. You should get sign-off from someone more authorative on IR like @nikic

mandlebug committed:
Address review comments and minor cleanup.

* Added support for emitting ref metadata on functions.
* Added testing for ref metadata on functions.
* Fixed up lang ref description.
* Renamed tests to match metadatas nameing.

mandlebug commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -766,6 +766,30 @@ void Verifier::visitGlobalValue(const GlobalValue &GV) {
+      SmallVector<MDNode *> MDs;
+      GO->getMetadata(LLVMContext::MD_ref, MDs);
+      for (const MDNode *MD : MDs) {
+        Check(MD->getNumOperands() == 1, "ref metadata must have one operand",
```
One difference with the callees metadata though is it is added in one place and never modified afterwards. For this metadata it is possible you could have 2 different features that require implicit references to be added. In that case would you have to get the existing array of refs, duplicate and append to it, then create a new metadata node with that array? Or are you suggesting we support both arrays of refs, and multiple individual refs as well? I'm not sure I'm particularly fond of either approach.

w2yehia replied:
>  For this metadata it is possible you could have 2 different features that require implicit references to be added. In that case would you have to get the existing array of refs, duplicate and append to it, then create a new metadata node with that array? 

yes you cannot mutate an existing node, so you'll have to create a new one, but that's not too much work as we constantly create new IR and throw old one in LLVM, and it will not happen often.

> Or are you suggesting we support both arrays of refs, and multiple individual refs as well?
No, not both because that would complicate the handling on the consumer side.
I'm in favor of supporting only the array of refs (and enforcing one `!ref` MD) just because it's more compact in textual IR (and I cannot think of other benefits).

mandlebug committed:
Add object file testing of ref reloc in functions.

mandlebug committed:
Verify ref metadata node is not placed on a declaration.

mandlebug committed:
Rename metadata from ref to implicit.ref.

mandlebug committed:
Propagate implicit.ref metadata when inlining.

Propagate the metadata from callee to caller on inlining. Also adds a
test verifying that the metadata is maintained when cloning a function.

hubert-reinterpretcast commented:
@mandlebug, this now needs rebase as it conflicts with the `alloc_token` metadata.

mandlebug committed:
Add implicit.ref metadata to the execptions that cannot be dropped.
